### DISCUSSION TITLE
Multiple &execute in a single call

### DISCUSSION THREAD

**Discussion by profucius** [Taitava/obsidian-shellcommands#316] [Ideas]
Hi there, I am not sure if this is a bug, or if I am doing something wrong. I scoured the documentation before coming here, so hopefully I did not miss anything, but I'm glad to hear the answer either way.

I am trying to trigger multiple commands in sequential order, via an Obsidian URI. I have two Shell Commands setup and working individually, and the third is setting a custom variable.

This is my URI:
obsidian://shell-commands/?vault=Vault&_customvariable=variable&execute=dksnzeqhno&execute=6wyzqh5q76

The reason I need it to be sequential is that the next depends on the previous to proceed. I click a button with the above URI, which first sets a custom variable, then executes another shell command which writes a text file of that variable name (effectively making it a global variable, so other applications can read it), and then after that is done it executes another shell command which contains a windows cmd to trigger an external application to read that filename/variable created previously. That external application then performs its task based on the variable passed to it.

Each one of these steps individually works fine, and if I combine steps 1 and 2 they work fine, but combining all three only sets the variable, but both executes do nothing. I suspect this is because it cannot perform two executes, because your plugin can only handle one at a time?

Again, not sure if this is a bug or just user error. If it's not a bug, then if you would be kind enough to tell me what I could do instead to get this working, that would be great! Technically, my sticking point here is making a shell-command variable become "global" so that other applications can read it. If there is a way to write to Windows Environmental Variables / Path, that would be amazing. Anything like that works. Thanks so much.

**Comment by Taitava** [comment]
Hi and thank you for your post! 🙂

Like you suspected, it's not possible to execute multiple shell commands via a single URI link. I have to admit that I didn't come to think about that possibility when I designed the feature. The idea is very clever and definitely something I want to implement! 😎

**After the implementation**, this is how your URI could look like: `obsidian://shell-commands/?vault=Vault&_customvariable=variable&execute=dksnzeqhno,6wyzqh5q76`

So, `execute=` would be present only once, and multiple shell command ids would be delimited by commas `,`.

The sequential order is a good idea and can be done. After executing the first shell command, the URI function could wait until it's finished and then execute the next shell command.

I'll move this discussion to _Ideas_. 🙂👍 

> Each one of these steps individually works fine, and if I combine steps 1 and 2 they work fine, but combining all three only sets the variable, but both executes do nothing. I suspect this is because it cannot perform two executes, because your plugin can only handle one at a time?

A general thing with any URIs (be it Obsidian or a website or whatever) is that they cannot contain the same key multiple times. When Obsidian sees two `execute=` keys in `obsidian://shell-commands/?vault=Vault&_customvariable=variable&execute=dksnzeqhno&execute=6wyzqh5q76`, it will only end up using **the latter one**: `execute=6wyzqh5q76`. The first one, `execute=dksnzeqhno`, is simply ignored. Generally, URIs can handle multiple same-named keys if they are postfixed with `[]`, so `obsidian://shell-commands/?vault=Vault&_customvariable=variable&execute[]=dksnzeqhno&execute[]=6wyzqh5q76` should be theoretically valid, but Obsidian does not seem to support `[]` (according to testing I just did on Obsidian `1.1.9`). (And even if it did support it, my plugin is not (yet) prepared to receive multiple shell command ids via a single URI link.)

Summary: Yes, only one can be executed at a time, and it's the last one in the URI that is executed. The first one goes totally unnoticed. In the future, `execute=dksnzeqhno,6wyzqh5q76` will be the syntax how it will work some day.

# A temporary workaround

> if you would be kind enough to tell me what I could do instead to get this working, that would be great!

I don't know exactly the contents of your shell commands, but generally you should be able to combine them into a single shell command. In case you didn't know, you can press Enter in the shell command field to enlarge it so each of the lines can contain one shell command. But if I understood correctly, you are using CMD.EXE as your shell? In that case, actually don't put the commands on separate lines. Put them on a single line and use `&` between the commands. For some reason, [when this plugin passes shell commands to CMD.EXE, linebreaks do not work as separators between commands](https://github.com/Taitava/obsidian-shellcommands/issues/306).

> If there is a way to write to Windows Environmental Variables / Path, that would be amazing.

I'm not certain if these will help you, but:
- [Additions to `PATH` can be done in the plugin's settings](https://publish.obsidian.md/shellcommands/Environments/Additions+to+the+PATH+environment+variable).
- Also, you could alter `PATH` and other environment variables in the actual shell command content:
  - [Use `SET` for setting environment variables temporarily](https://superuser.com/a/79614/1606938)
  - [Use `SETX` for setting environment variables permanently](https://stackoverflow.com/a/5898197/2754026)
  - Both of these assume you are using CMD.EXE as your shell.

**Reply by Taitava** [reply]
## I renamed the discussion
Old title: Multiple &execute in a single call, fails
New title: Multiple &execute in a single call

**Comment by profucius** [comment]
Thank you so much for the helpful reply! This all makes sense to me. After giving your post some careful consideration and doing some research on CMD and PowerShell, I've decided to switch to PowerShell commands within your plugin. Also, while testing the SETX command, I serendipitously found a solution to passing the variable directly to my external application (macro automation software) through the PowerShell command. This means I can run a single call with a custom variable and a single &execute to do everything I need! This is truly amazing!

I'm still looking forward to whenever you're able to implement multiple executes, as there may be another need for that in the future. But for now, you've helped point me in the right directly to resolve all my needs. Thanks again for being so helpful, and of course writing such a powerful plugin. Cheers!

**Reply by Taitava** [reply]
Thank you for your kind feedback! 🌞

I'm glad you were able to solve this particular multi-command execution case. 👍 

Switching from CMD to PowerShell has some good things from this plugin's side, too:
 - Variables are safer to use, as special characters in their [values are escaped](https://publish.obsidian.md/shellcommands/Variables/Escaping+special+characters+in+variable+values).
 - Multiline commands work properly, as opposed to: #306

But there are a couple of negative things, too:
 - PowerShell is slower to start, so each command has a one or two seconds delay.
 - ~~[Error messages might look crappy](https://github.com/Taitava/obsidian-shellcommands/discussions/155#discussioncomment-2194856)~~ (edit 2024-04-29: Fixed by #329), and there are some [encoding problems](https://github.com/Taitava/obsidian-shellcommands/discussions/179) in case you are using non-English chracters. These two are probably mistakes on this plugin's side, so I'm not blaming PowerShell 🙂. I just haven't used enough time to troubleshoot them yet.

**Comment by Taitava** [comment]
# Syntax for executing multiple shell commands

> In the future, execute=dksnzeqhno,6wyzqh5q76 will be the syntax how it will work some day.

I think this could be extended by allowing different separators between shell commands:
- Comma `,`: Execute the latter shell command after the former regardless of the result of the first one.
- Plus `+`: Execute the latter shell command only if the former succeeded*.
- Minus `-`: Execute the latter shell command only if the former failed*.
- Semicolon `;`: Define another group of shell commands that executes at the same time with the former group.

Example: `?execute=commandA1+commandA2-commandA3,commandA4;commandB1,commandB2`
- `commandA1` gets always executed.
  - `commandA2` gets executed only if A1 succeeded.
  - `commandA3` gets executed only if A1 failed, or if A2 failed.
- `commandA4` gets always executed, but only after A2 **or** A3 have finished executing.
- `commandB1` gets always executed, and it starts the same time with `commandA1`.
  - `commandB2` gets executed after B1, regardless of B1's result.

**This is just future planning, none of this is implemented yet.** The first version will probably only have comma `,`.

I was thinking about Bash style operators `&&` (= and) and `||` (= or), but `&&` would require url encoding, so I'm ruling them out to make it easy to handcraft uris.

*) _Succeeding_ means that a shell command returns an exit code `0`, or [an exit code that is ignored](https://publish.obsidian.md/shellcommands/Output+handling/Ignoring+errors#Ignore+error+messages+based+on+a+shell+command's+exit+code).
